rskj-core/src/main/java/co/rsk/logfilter/BlocksBloomStore.java

```diff
@@ -29,22 +32,58 @@
     private final int noBlocks;
     private final int noConfirmations;
     private final Map<Long, BlocksBloom> blocksBloom = new ConcurrentHashMap<>();
+    private final KeyValueDataSource dataSource;
 
-    public BlocksBloomStore(int noBlocks, int noConfirmations) {
+    public BlocksBloomStore(int noBlocks, int noConfirmations, KeyValueDataSource dataSource) {
         this.noBlocks = noBlocks;
         this.noConfirmations = noConfirmations;
+        this.dataSource = dataSource;
     }
 
     public boolean hasBlockNumber(long blockNumber) {
-        return this.blocksBloom.containsKey(this.firstNumberInRange(blockNumber));
+        if (this.blocksBloom.containsKey(this.firstNumberInRange(blockNumber))) {
+            return true;
+        }
+
+        if (this.dataSource != null && this.dataSource.get(longToKey(blockNumber)) != null) {
+            return true;
+        }
+
+        return false;
     }
 
     public BlocksBloom getBlocksBloomByNumber(long number) {
-        return this.blocksBloom.get(firstNumberInRange(number));
+        long key = firstNumberInRange(number);
+
+        BlocksBloom blocksBloom = this.blocksBloom.get(key);
+
+        if (blocksBloom != null) {
+            return blocksBloom;
+        }
+
+        if (this.dataSource == null) {
+            return null;
+        }
+
+        byte[] data = this.dataSource.get(longToKey(key));
```

Comment:
wouldn't you want to load everything in memory when it starts?

Reply:
We don't have the information at start (the log blooms are built on demand). The node maybe never would be asked for that log range, etc... So the memory cache is only to working nodes that have a behavior: someone is requesting the logs, from JSON RPC. The store is the new piece in this pull request: then, when a node restart and is queried for logs, it could respond faster.